README.md
Sample Document
================



















<!-- README.md is generated from README.Rmd. Please edit that file -->

# pCODE

<!-- badges: start -->

<!-- badges: end -->

The `R` package `pCODE` offers more user-friendly functions for
estimating ODE models without specifying any derivatives. `pCODE` also
includes a bootstrap variance estimator in addition to the estimator
obtained by Delta method. `pCODE` uses k-fold cross-validation for
choosing an optimal penalty parameter. The estimation procedure follows
Ramsay et al. (2007) which presents a new approximation strategy in the
family of collocation methods. It combines data smoothing with
generalized profiling algorithm to estimate parameters of an ODE model
where the solutions are subsequently obtained upon the parameter
estimates.

## Installation

For now, the developing package can be installed
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github('alex-haixuw/PCODE')
```

The package is being submitted to \[CRAN\], and later you can install
the released version of PCODE from [CRAN](https://CRAN.R-project.org)
with:

``` r
install.packages("pCODE")
```

## Example

A simple illustration uses an one-dimensional ODE model \[
\dot{X} = \theta X (1-\frac{X}{10})
\] The following code defines the forementioned model that will be
provided for `pcode` for estimating parameters and `ode` for obtaining
numeric solution:

``` r
#load dependencies
library(pCODE)
library(deSolve)
library(fda)
library(MASS)
library(pracma)
library(Hmisc)
#set seed for reproducibility
set.seed(123)
ode.model <- function(t,state,parameters){
            with(as.list(c(state,parameters)),{
                 dX <- theta*X*(1-X/10)
                return(list(dX))})}
```

Let \(\theta = 0.1\) and \(X(0) = 0.1\)

``` r
#define model parameters
model.par   <- c(theta = c(0.1))
#define state initial value
state       <- c(X     = 0.1)
```

Given an observation period of \([0,100]\), random noise errors are
added to the ODE solution with a Normal distribution
\(\text{N}(0,0.5^{2})\). Observations are generated as follows:

``` r
times  <- seq(0,100,length.out=101)
mod    <- ode(y=state,times=times,func=ode.model,parms = model.par)
nobs   <- length(times)
scale  <- 0.5
noise  <- scale * rnorm(n = nobs, mean = 0 , sd = 1)
observ <- mod[,2] + noise
```

Subsequently, we can visualize the observations along the true solution
of this simple ODE model as

``` r
#plot simulated data against generating model
plot(mod,ylab=names(state))      #curve
points(times, observ,pch='*',col='blue')    #observation
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" style="display: block; margin: auto;" />
First, a basis object needs to be defined by `create.bspline.basis` from
`fda` package for smoothing observations. A B-spline basis is created
given 21 konts including both interior and boundary knots acorss
observation interval \((0,100)\). The order of basis functions is 4, so
there is a total of 23 basis functions.

``` r
#Generate basis object for interpolation and as argument of pcode
#21 konts equally spaced within [0,100]
knots <- seq(0,100,length.out=21)
#order of basis functions
norder <- 4
#number of basis funtions
nbasis <- length(knots) + norder - 2
#creating Bspline basis
basis  <- create.bspline.basis(c(0,100),nbasis,norder,breaks = knots)
```

To perform parameter cascade method for estimating both structural and
nuisance parameters, one can use `pcode` in the following way

``` r
#parameter estimation
pcode.result <- pcode(data = observ, time = times, ode.model = ode.model,
                      par.initial = 0.3, par.names = 'theta',state.names = 'X',
                      basis.list = basis, lambda = 1e2)
```

The structural parameter and nuisance parameter estiamtes can be called
by

``` r
pcode.result$structural.par
#>      theta 
#> 0.09995229
pcode.result$nuisance.par
#>  [1]  0.1232160  0.1550332  0.1903906  0.2729993  0.4284428  0.6134020
#>  [7]  1.0222183  1.6891098  2.5351231  3.5543293  4.8116926  6.0699739
#> [13]  7.2145361  8.0734588  8.7456720  9.2110485  9.4866269  9.7101657
#> [19]  9.8736650  9.9650449 10.0098433  9.9950749  9.9846698
```

## References

<div id="refs" class="references">

<div id="ref-parcascade">

Ramsay, J. O., G. Hooker, D. Campbell, and J. Cao. 2007. “Parameter
Estimation for Differential Equations: A Generalized Smoothing
Approach.” *Journal of the Royal Statistical Society. Series B
(Statistical Methodology)* 69 (5): 741–96.

</div>

</div>


NEWS.md
## pCODE 0.9.4
* Add verval options
* Add automated tests for package functions

## pCODE 0.9.3
* Fixed likelihood problem
* Update license information

## pCODE 0.9.2
* Fixed verbal problem

## pCODE 0.9.1

* Initial release


tests/testthat.R
library(testthat)
library(pCODE)
library(deSolve)
library(fda)
library(MASS)
library(pracma)
library(rmarkdown)

test_check("pCODE")


simulation_test.R
rm(list=ls())

library(deSolve)
library(fda)
library(MASS)
library(R.matlab)
library(pracma)
library(tictoc)
source('par_cascade_functions.R')

#####
#Define ode model 
model.par   <- c(theta = c(0.1))
state       <- c(X     = 0.1)

ode.model <- function(t, state,parameters){
  with(as.list(c(state,parameters)),
       {
         dX <- theta*X*(1-X/10)
         return(list(dX))
       })
}

Dmodel <- function(state,parameters){
  with(as.list(c(state,parameters)),
       {
         dX <- theta*X*(1-X/10)
         return(list(dX))
       })
}

#Observation points
times <- seq(0,100,length.out=101)

desolve.mod <- ode(y=state,times=times,func=ode.model,parms = model.par)


#Plot solution
plot(desolve.mod,main='noisy data') #curve
points(times,desolve.mod[,2],col='red') #Points

theta.est.optim <- cbind(rep(NA,100),rep(NA,100))

for (jj in 1:100){
  print(jj)
#Simulate data
nobs  <- length(times)
scale <- 0.5
noise <- scale*rnorm(n = nobs, mean = 0 , sd = 1)

observ <- desolve.mod[,2] + noise
#plot simulated data against generating model
points(times, observ,pch='*',col='blue')


#--------------------------------------------------------
#Generating basis functions for interpolating observations


#knots located @ each observation time point
knots <- seq(0,100,length.out=21)
#order of basis functions
norder <- 5
#number of basis funtions
nbasis <- length(knots) + norder - 2
#creating basis
basis  <- create.bspline.basis(c(0,100),nbasis,norder,breaks = knots)


nls_temp = cascade_nls(0.3, ode.model = Dmodel, basis , observ, times,controls = list(tau = 0.1, tolx = 1e-8,tolg = 1e-8, maxeval = 500))

phi.mat <- eval.basis(times,basis)
temp <- phi.mat %*% nls_temp$nuisance.par 
points(times,temp,pch='^',col='green')




optim_temp = cascade_optim(0.1, ode.model = Dmodel, basis, observ, times,oned_range = c(0,1))




theta.est.optim[jj,1] <- nls_temp$structural.par
theta.est.optim[jj,2] <- optim_temp$structural.par

}

par(mfrow=c(1,2))
boxplot(theta.est.optim[,1],main='NLS')
abline(h=0.1,col='red')
boxplot(theta.est.optim[-39,2],main='OPTIM')
abline(h=0.1,col='red')
mtext("SD = 0.1", side = 3, line = -26, outer = TRUE)



#save(theta.est.optim, file = 'theta.est.optim_3.Rdata')
rm(list=ls())
theta.mat <- readMat('theta_est_nonlin_3.mat')
load('theta.est.optim_3.Rdata')
rm(list=ls())
theta.mat <- readMat('theta_est_lsqnonlin.mat')
load('theta.est.optim.Rdata')

theta_est_nonlin <- as.vector(theta.mat$theta.est)

type = c(rep('optim',1000),rep('lsqnonlin',1000))


temp <- c(theta.est.optim, -theta_est_nonlin)
data.mat = data.frame( theta.est = temp, est.type = type)
boxplot(data.mat$theta.est ~ data.mat$est.type)



tests/testthat/test-PCODE.R
context('pcode')
test_that("pcode works in parameter estimation",{
  set.seed(123)
  ode.model <- function(t,state,parameters){
    with(as.list(c(state,parameters)),{
      dX <- theta*X*(1-X/10)
      return(list(dX))})}
  #define model parameters
  model.par   <- c(theta = c(0.1))
  #define state initial value
  state       <- c(X     = 0.1)
  times  <- seq(0,100,length.out=101)
  mod    <- ode(y=state,times=times,func=ode.model,parms = model.par)
  nobs   <- length(times)
  scale  <- 0.5
  noise  <- scale * rnorm(n = nobs, mean = 0 , sd = 1)
  observ <- mod[,2] + noise
  #Generate basis object for interpolation and as argument of pcode
  #21 konts equally spaced within [0,100]
  knots <- seq(0,100,length.out=21)
  #order of basis functions
  norder <- 4
  #number of basis funtions
  nbasis <- length(knots) + norder - 2
  #creating Bspline basis
  basis  <- create.bspline.basis(c(0,100),nbasis,norder,breaks = knots)

  pcode.result <- pcode(data = observ, time = times, ode.model = ode.model,
                        par.initial = 0.1, par.names = 'theta',state.names = 'X',
                        basis.list = basis, lambda = 1e2)
  expect_equal( (abs(unname(pcode.result$structural.par) - unname(model.par))/abs(unname(model.par))) <= 1e-1, TRUE)
})


main_code.R
rm(list=ls())

library(deSolve)
library(fda)
library(MASS)
source('par_cascade_functions.R')

theta.true <- -5
x0.initial <- 10

#Derivative model
deriv_model <- function(t, y, parms){
  theta = parms
  list(theta*y)
}

#Observation points
times <- seq(0,1,length.out=21)

#Set up ODE system
ode.system <- ode(y = x0.initial, times = times, func = deriv_model,parms = theta.true, method='ode45')

#check values
head(ode.system, n = 5)
tru.obs <- ode.system[,2]


#Plot solution
plot(ode.system) #curve
points(times,ode.system[,2],col='red') #Points


#Simulate data
nobs  <- length(times)
scale <- 0.1
noise <- scale*rnorm(n = nobs, mean = 0 , sd = 1)

observ <- tru.obs + noise
#plot simulated data against generating model
points(times, observ,pch='*',col='blue')


#--------------------------------------------------------
#Generating basis functions for interpolating observations


#knots located @ each observation time point
knots <- times
#order of basis functions
norder <- 4
#number of basis funtions
nbasis <- length(knots) + norder - 2
#creating basis
basis  <- create.bspline.basis(c(0,1),nbasis,norder,breaks = knots)

cascade_est(-4, derive_model, basis, observ, times,oned_range = c(-6,-4))


multi_dim.R
#############################################################################
#
#
#
#
#
#############################################################################


#Function list
#1. prepare_basis
#2. PC_ODE
#3. innerobj_multi
#4. outterobj_multi_nls
#5. innerobj
#6. outterobj
#7. cascade_nls
#
#
#



###############
#1. Function : 'prepare_basis'
#
#        Given a basis object, it evaluates the basis over both observation points and equally spaced quadtuare points.
#
#Input:
#           x: A basis object
#       times: A vector containing Observation points
#    nquadpts: A numeric value that defines the number of quadrature points
#
#Output:
#     Phi.mat: The matrix contains evaluations of all basis functions at every observation times, where evaluations of each basis
#                    function are  stored as the columns of the matrix.
#        Qmat: The matrix contains evaluations of all basis functions at every quadrature points in the support, where evaluations of
#                    each basis function are stored as the columns of the matrix.
#     Q.D1mat: The matrix contains evaluations of the first order derivative of all basis functions at every quadature points.
#     Q.D2mat: The matrix contains evaluations of the second order derivative of all basis functions at every quadature points.
#      quadts: A vector of quadrature points.
#     quadwts: A vector of quadrature weights for each quadrature points.
#
#Comment:
#
#
###############
prepare_basis <- function(x, times, nquadpts){

  #Evaluate basis functions over observation time points
  Phi.mat <- eval.basis(times, x)

  #Preparation to calculate L2 penalty
  #Evaluate basis function over quadrature points, and the number of
  #quadrature points, 'nquadpts', defined the density of points.
  quadts   <- seq(min(times),max(times),length.out=nquadpts)
  nquad    <- length(quadts)
  quadwts  <- rep(1,nquad)
  even.ind <- seq(2,(nquad-1),by=2)
  odd.ind  <- seq(3,(nquad-2),by=2)
  quadwts[even.ind] = 4
  quadwts[odd.ind] = 2
  h  <- quadts[2] - quadts[1]
  quadwts <- quadwts*(h/3)

  Qmat    <- eval.basis(quadts, basis)
  Q.D1mat <- eval.basis(quadts, basis, 1)
  Q.D2mat <- eval.basis(quadts, basis, 2)

  return(list(Phi.mat = Phi.mat, Qmat = Qmat, Q.D1mat = Q.D1mat, Q.D2mat = Q.D2mat, quadts= quadts, quadwts = quadwts))

}


###############
#2. Function: 'PC_ODE'
#
#
#Input:
#
#         data: A data frame or a matrix contain observations from each dimension of the ODE model.
#         time: A vector contain observation times or a matrix if time points are different between dimensions.
#    ode.model: A function that computes the time derivative of the ODE model given states variable evaluated at a given time.
#    par.names: A vector contains the names of structural parameters defined in the 'ode.model'.
#  state.names: A vector contains the names of state variables defined in the 'ode.model'.
#  par.initial: Initial value of structural parameters to be optimized.
#   basis.list: A list of basis objects for smoothing each dimension's observations. Can be the same or different across dimensions.
#       lambda: Penalty parameter.
#     controls: A list of control parameters. See ‘Details’.
#
#Output:
#
#   structural.par: The structural parameters of the ODE model.
#    nuissance.par: The nuissance parameters or the basis coefficients for interpolating observations.
#
#
#Comment:
#
###############
#data, time , ode.model, state.names, par.names, par.initial = NULL, basis.list = DEFAULT (order = 5,
PC_ODE <- function(data, time, ode.model,par.names,state.names,  par.initial, basis.list, lambda = NULL,controls = NULL){

      #Set up default controls for optimizations and quadrature evaluation
      con.default <- list(nquadpts = 101, smooth.lambda = 1e5, tau = 0.01, tolx = 1e-6,tolg = 1e-6, maxeval = 30)
      #Replace default with user's input
      con.default[(namc <- names(controls))] <- controls
      con.now  <- con.default


      #Evaluate basis functiosn for each state variable
      basis.eval.list <- lapply(basis.list, prepare_basis, times = time, nquadpts = con.now$nquadpts)

      #Number of dimensions of the ODE model
      ndim <- ncol(data)

      #Some initializations
      inner.input  <- list()
      initial_coef <- list()

      for (ii in 1:ndim){
        #For each dimension, obtain initial value for the nuissance parameters or the basis coefficients
        Rmat = t(basis.eval.list[[ii]]$Q.D2mat)%*%(basis.eval.list[[ii]]$Q.D2mat*(basis.eval.list[[ii]]$quadwts%*%t(rep(1,basis.list[[ii]]$nbasis))))
        basismat2 = t(basis.eval.list[[ii]]$Phi.mat)%*%basis.eval.list[[ii]]$Phi.mat;
        Bmat    = basismat2 + con.default$smooth_lambda*Rmat;
        #Initial basis coefficients
        initial_coef[[ii]] = ginv(Bmat)%*%t(basis.eval.list[[ii]]$Phi.mat)%*%data[,ii]
        inner.input[[ii]]  = list(data[,ii], basis.eval.list[[ii]]$Phi.mat, lambda,
                                  basis.eval.list[[ii]]$Qmat, basis.eval.list[[ii]]$Q.D1mat,
                                  basis.eval.list[[ii]]$quadts, basis.eval.list[[ii]]$quadwts,time,
                                  state.names,model.names)
      }


      #Searching for a better starting value for optimization
      #temp <- lsqnonlin(innerobj_multi, unlist(initial_coef), ode.par = par.initial, derive.model = ode.model, input = inner.input, NLS=TRUE,options = list(tau = 0.01))$x

      #Running optimization for outter objective function to obtain structural parameter
      par.final <- lsqnonlin(options = list(maxeval = con.default$maxeval,tau = con.default$tau),outterobj_multi_nls, par.initial,basis.initial = unlist(initial_coef), derivative.model = ode.model, inner.input = inner.input)$x
      #Condition on the obtained structural parameter, calculate the nuissance parameter or the basis coefficients to interpolate data
      basis.coef.final <- lsqnonlin(innerobj_multi, unlist(initial_coef), ode.par = par.final, derive.model = ode.model, input = inner.input, NLS=TRUE,options = list(tau =0.01))$x


      return(list(structural.par = par.final, nuissance.par = basis.coef.final))
}


#####
#'innerobj_multi':
#
#            Conditioning on the structural parameters of the ODE model, 'ode.par', this functions returns either the
#            inner objective function value (the sum of squared errors) or the residual between the observation and
#            and the basis expansion estimates. The form of returning value depends on the later called optimizers.
#
#
#Input:
#     basis_coef: A single vector containing the basis coefficients of all dimensions of the ODE model.
#        ode.par: A vector or a numeric value containing the structural parameter of the ODE model.
#   derive.model: The function defines the time derivative of of each state variable
#           NLS : A logic value determines whether the function should return a numeric value (NLS = FALSE)
#                                                                    or a vector of residuals (NLS = TRUE)
#Output
#
#         residual.vec : A vector containing the residuals from using basis expansion to estimate observations
#                        and the quarature evaluations to approximate the ODE penalty,
#   sum(residual.vec^2): A numeric value correspondes to the inner objective function value.
#
#Comment:
#
#
#
innerobj_multi  <- function(basis_coef, ode.par, input, derive.model,NLS=TRUE){


    #Get variables  from 'input'
    ndim     <- length(input)
    npoints  <- length(unlist(input[[1]][8]))
    nbasis   <- rep(NA, ndim+1)
    Xhat     <- matrix(NA, nrow = npoints, ncol = ndim)
    residual <- matrix(NA, nrow = npoints, ncol = ndim)

    state.names <- input[[1]][[9]]
    model.names <- input[[1]][[10]]
    Xt   <- matrix(NA, nrow = length(input[[1]][[7]]),ncol= ndim)
    dXdt <- matrix(NA, nrow = length(input[[1]][[7]]),ncol= ndim)

    #Turn a single vector 'basis_coef' into seperate locations
    coef.list <- list()
    nbasis[1] <- 0
    for (jj in 1:ndim){
       nbasis[jj+1]    <- ncol(input[[jj]][[2]])
       coef.list[[jj]] <- basis_coef[(nbasis[jj]+1):(nbasis[jj]+nbasis[jj+1])]
       #Basis expansion for j-th dimension
       Xhat[,jj]       <- input[[jj]][[2]] %*% coef.list[[jj]]
       #Residual from basis expansion
       residual[,jj]   <- input[[jj]][[1]] - Xhat[,jj]
       Xt[,jj]         <- input[[jj]][[4]] %*% coef.list[[jj]]
       dXdt[,jj]       <- input[[jj]][[5]] %*% coef.list[[jj]]
    }

   #Ode penalty


   names(ode.par) <- model.names

   temp_fun  <- function(x, temp = ode.par,names = state.names){
       names(x) <- state.names
       return(unlist(derive.model(state = x, parameters = temp)))
   }
   temp_eval <- t(apply(Xt,1,temp_fun))

   temp_list <- list(dXdt, temp_eval)
   temp_penalty_resid <- Reduce("-", temp_list)

   lambda <- input[[1]][[3]]
   penalty_residual <- matrix(NA, nrow = nrow(temp_eval),ncol=ncol(temp_eval))
   for (jj in 1:ndim){
     penalty_residual[,jj] <- sqrt(lambda) * sqrt(input[[jj]][[7]]) * temp_penalty_resid[,jj]
   }




   residual.vec <- c(as.vector(residual), as.vector(penalty_residual))
     if (NLS){
       return(residual.vec)
     }else{
       return(sum(residual.vec^2))
     }

}

#####
#
#####
outterobj_multi_nls <- function(ode.parameter, basis.initial, derivative.model, inner.input,NLS=TRUE){
  #Convergence of basis coefficients seems to happen before 'maxeval'.

  inner_coef <- lsqnonlin(innerobj_multi, basis.initial, ode.par = ode.parameter, derive.model = derivative.model, options = list(maxeval = 100,tolx=1e-6,tolg=1e-6), input = inner.input)$x
  #inner_coef <- optim(basis.initial,innerobj_multi,  ode.par = ode.parameter, derive.model = derivative.model, input = inner.input,NLS=FALSE)$par
  #Get variables  from 'input'
  ndim     <- length(inner.input)
  npoints  <- length(inner.input[[1]][[8]])
  basisnumber   <- rep(NA, ndim+1)
  Xhat     <- matrix(NA, nrow = npoints, ncol = ndim)
  residual <- matrix(NA, nrow = npoints, ncol = ndim)
  #Turn a single vector 'inner_coef' into seperate locations
  coef.list <- list()
  basisnumber[1] <- 0
  for (jj in 1:ndim){
    basisnumber[jj+1]    <- ncol(inner.input[[jj]][[2]])
     coef.list[[jj]] <- inner_coef[(basisnumber[jj]+1):(basisnumber[jj]+basisnumber[jj+1])]
     #Basis expansion for j-th dimension
     Xhat[,jj]       <- inner.input[[jj]][[2]] %*% coef.list[[jj]]
     #Residual from basis expansion
     residual[,jj]   <- inner.input[[jj]][[1]] - Xhat[,jj]
  }

  if(NLS){
    return(as.vector(residual))
  }else{
    return(sum(residual^2))
  }


}





#####
#  'innerobj'
#
# Description:
#
#   'innerobj' is the inner objective function to be optimzied for estimating nuisance
#    parameters, the basis coefficients, of the model when a new structural parameter
#    estimate is available.
#
#    This objective function contains two component:
#        1. 'SSE' corresponds to the sum of squared errors between the observations
#                and the estiamted trajectory by basis expansion.
#        2. 'ode_penalty' controls the size of the extend how the estimated trajectory
#                fails to satisfy the differential equation exactly given a smoothing
#                parameter lambda.
#
# Input:
#      basis_coef      :
#      ode.par         :
#      input           :
#      derive.model    :
#
# Output:
#
#     SSE + ode_penalty :
#     residual.vec      :
#
# Comment:
#
#
#
#
#####
innerobj <- function(basis_coef, ode.par, input, derive.model,NLS=TRUE){
  yobs    <- input[[1]]
  Phi.mat <- input[[2]]
  lambda  <- input[[3]]
  Qmat    <- input[[4]]
  Q.D1mat <- input[[5]]
  quadts  <- input[[6]]
  quadwts <- input[[7]]

  xhat     <- Phi.mat %*% basis_coef
  #e_{ij}
  residual <- yobs - xhat
  #Sum of squared errors
  #SSE <- sum((yobs-xhat)^2)

  #PEN_{i}
  penalty_residual <- rep(NA, length(quadwts))
  #Use composite Simpson's rule to calculate the integral of residual
  Xt   <- Qmat    %*%  basis_coef
  dXdt <- Q.D1mat %*%  basis_coef

  #Evaluate f(X)
  rightside <- rep(NA, length(quadwts))

  tempfun <- function(x, temp = ode.par){
    return(derive.model(state = c(X=x), parameters= c(theta = temp)))
  }
  rightside <- unlist(lapply(Xt,tempfun))

  #for (jj in 1:length(quadwts)){
  #      rightside[jj] <- derive.model(state = c(X = Xt[jj]),parameters = c(theta = #ode.par))
  #}

  penalty_residual <- sqrt(lambda) * sqrt(quadwts) * (dXdt - rightside)

  #ode_penalty <- sum(lambda * quadwts * (penalty_residual)^2)

  residual.vec <- c(residual,penalty_residual)
  if (NLS){
    return(residual.vec)
  }else{
    return(sum(residual.vec^2))
  }
}


outterobj <- function(ode.parameter, basis.initial, derivative.model, inner.input,NLS){


  if (NLS){
    inner_coef  <- lsqnonlin(innerobj, basis.initial, options=list(maxeval = 50),ode.par = ode.parameter,derive.model = derivative.model, input = inner.input)$x
  }else{
    inner_coef <- optim(par=basis.initial, fn=innerobj,ode.par = ode.parameter, derive.model = derivative.model, input = inner.input,NLS = FALSE,
                        control = list(abstol=1e-10))$par
  }



  yobs    <- inner.input[[1]]
  Phi.mat <- inner.input[[2]]


  xhat     <- Phi.mat %*% inner_coef
  #e_{ij}
  residual <- yobs - xhat
  if (NLS){
    return(residual)
  }else{
    return(sum(residual^2))
  }

}

cascade_nls <- function(par.initial, ode.model, basis, observations, obs.times,ode_penalty = 1e6,nquadpts = NULL,controls = NULL){

    #Input:
    #    par.initial : initial values for structural parameter
    #    ode.model   : ODE model
    #    basis       : basis object for approximating x(t)
    #    observations: observations including a vector indicating times of observations
    #    nquadpts    : number of quadrature points to use composite Simpson's rule for calculating integral


    #Output:



    #Comment:

    #number of parameters
    npar  <- length(par.initial)
    #time points of observations


    #If no 'quadpts' is given, then set it to 101
    if (is.null(nquadpts)){
        nquadpts = 101
    }



    #Evaluating basis functions at time points of observations
    #and stored as columns in Phi matrix
    Phi.mat <- eval.basis(obs.times, basis)
    #Evaluating 1st derivative of basis functions
    D1.mat <- eval.basis(obs.times, basis,1)
    #Evaluating 2nd derivative of basis functions
    D2.mat <- eval.basis(obs.times, basis,2)

    #Calculate L2 penalty
    quadts   <- seq(min(obs.times),max(obs.times),length.out=nquadpts)
    nquad    <- length(quadts)
    quadwts  <- rep(1,nquad)
    even.ind <- seq(2,(nquad-1),by=2)
    odd.ind  <- seq(3,(nquad-2),by=2)
    quadwts[even.ind] = 4
    quadwts[odd.ind] = 2
    h  <- quadts[2] - quadts[1]
    quadwts <- quadwts*(h/3)

    Qmat    <- eval.basis(quadts, basis)
    Q.D1mat <- eval.basis(quadts, basis, 1)
    Q.D2mat <- eval.basis(quadts, basis, 2)

    #Initial estimat of basis coefficients
    Rmat = t(Q.D2mat)%*%(Q.D2mat*(quadwts%*%t(rep(1,nbasis))))
    basismat2 = t(Phi.mat)%*%Phi.mat;
    smooth_lambda = 1e-6   # smoothing parameter
    Bmat    = basismat2 + smooth_lambda*Rmat;
    #Initial basis coefficients
    initial_coef = ginv(Bmat)%*%t(Phi.mat)%*%observations

    #Passing to inner objective functions and obtain initial value for parameter cascading
    inner.input = list(observations, Phi.mat, 1e5, Qmat, Q.D1mat,quadts,quadwts)



      #Using nonlinear least square for optimization
      temp <- lsqnonlin(innerobj, initial_coef, ode.par = par.initial, derive.model = ode.model, input = inner.input,NLS = TRUE)
      new.ini.basiscoef <- matrix(temp$x,length(temp$x),1)
      #--------------------------------------------------------

      theta.final  <- lsqnonlin(outterobj, par.initial, options = controls,basis.initial = initial_coef, derivative.model = ode.model,inner.input = inner.input,NLS=TRUE)$x
      basiscoef <- lsqnonlin(innerobj, initial_coef, ode.par = theta.final, derive.model = ode.model, input = inner.input,NLS = TRUE)$x

    return(list(structural.par = theta.final, nuisance.par = basiscoef))
}


examples/FitzHugh_example.R
rm(list=ls())

library(deSolve)
library(fda)
library(MASS)
library(pracma)
source('/Users/Inori/Desktop/PCODE/R/PCODE.R')

#####

#Define ode model
model.par   <- c(a=0.2,b=0.2,c=3)
state       <- c(V=-1,R=1)

ode.model <- function(t, state,parameters){
  with(as.list(c(state,parameters)),
       {
         dV <- c*(V - (V^3)/3 + R)
         dR <- -(1/c) * (V - a + b*R)
         return(list(c(dV,dR)))
       })
}

Dmodel <- function(state,parameters){
  with(as.list(c(state,parameters)),
       {
         dV <- c*(V - (V^3)/3 + R)
         dR <- -(1/c) * (V - a + b*R)
         return(list(c(dV,dR)))
       })
}
#Observation points
times <- seq(0,20,length.out=401)
#Generate ODE
desolve.mod <- ode(y=state,times=times,func=ode.model,parms = model.par)
#
par(mfrow=c(2,1))
par(mar = c(4,4,2,2) + 0.1)
plot(desolve.mod[,1],desolve.mod[,2],type='l',main='FitzHugh-Nagumo model',ylab='V(t)',xlab='',lwd=2)
plot(desolve.mod[,1],desolve.mod[,3],type='l',main='',ylab='R(t)',xlab='t',lwd=2)


#simulations
nobs  <- length(times)
scale <- 0.1
noise <- scale*rnorm(n = nobs, mean = 0 , sd = 1)

observ <- matrix(NA, nrow = length(times),ncol =3)
observ[,1] <- times
observ[,2] <- desolve.mod[,2] + noise
observ[,3] <- desolve.mod[,3] + noise
#------------------------------------------------
plot(desolve.mod[,1],desolve.mod[,2],type='l',main='V(t)',ylab='V(t)',xlab='t')
points(desolve.mod[,1],desolve.mod[,2],col='red')
points(desolve.mod[,1],observ[,2],col='blue')
#------------------------------------------------
plot(desolve.mod[,1],desolve.mod[,3],type='l',main='R(t)',ylab='R(t)',xlab='t')
points(desolve.mod[,1],desolve.mod[,3],col='red')
points(desolve.mod[,1],observ[,3],col='blue')
#------------------------------------------------


#Define basis
knots <- seq(0,20,length.out=101)
#order of basis functions
norder <- 4
#number of basis funtions
nbasis <- length(knots) + norder - 2
#creating basis
basis  <- create.bspline.basis(c(0,20),nbasis,norder,breaks = knots)

basis.list <- list(basis,basis)
data <- observ[,2:3]
colnames(data) <- c('V','R')


#1. Running parameter cascade
pcode.result <- pcode(data = data, time= observ[,1], ode.model = Dmodel, par.names = c('a','b','c'),
                     state.names = c('V','R'), par.initial = c(0.1,0.3,4),lambda = 1e2,basis.list = basis.list,
                     controls = list(smooth.lambda = 1e-3,verbal = 1,maxeval = 20))
pcode.result$structural.par



#2.
deltavar.res <- deltavar(data = observ[,2:3], time= observ[,1], ode.model = ode.model, par.names = c('a','b','c'),
                               state.names = c('V','R'), par.initial = c(0.1,0.3,4),lambda = 1e2,basis.list = basis.list,
                          controls = list(smooth.lambda = 10,verbal = 1,maxiter = 50, inner.maxeval = 30),stepsize = 0.001, y_stepsize =0.001)
deltavar.res

#3.
bootsvar.res <- bootsvar(data = observ[,2:3], time= observ[,1], ode.model = ode.model, par.names = c('a','b','c'),
                         state.names = c('V','R'), par.initial = rnorm(3),lambda = 1e2,basis.list = basis.list,
                         controls = list(smooth.lambda = 10,verbal = 1,maxiter = 50, inner.maxeval = 30),bootsrep = 20)
bootsvar.res
#4.
lambda.check <- tunelambda(data =  observ[,2:3], time = observ[,1], ode.model = ode.model, par.initial = c(a=0.2,b=0.2,c=3),
                    par.names = c('a','b','c'), state.names = c('V','R'),basis.list = basis.list, lambda_grid = 10^(-2:3),cv_portion = .05,
                    rep = 2, kfolds = 5, controls = list(smooth.lambda = 10,verbal = 1,maxiter = 50, inner.maxeval = 30))
lambda.check
